components/ai_chat/content/browser/ai_chat_tab_helper.cc

```diff
@@ -321,52 +278,105 @@ void AIChatTabHelper::GetPageContent(GetPageContentCallback callback,
-  } else {
-    if (base::Contains(kPrintPreviewRetrievalHosts,
-                       GetPageURL().host_piece())) {
-      pending_print_preview_requests_[pending_navigation_id_] =
-          std::move(callback);
-      if (is_page_loaded_) {
-        NotifyPrintPreviewRequested(false);
-      }
+    return;
+  }
+  if (base::Contains(kPrintPreviewRetrievalHosts, GetPageURL().host_piece()) &&
+      print_preview_extraction_delegate_ != nullptr) {
+    if (is_page_loaded_) {
+      // Get content using a printing / OCR mechanism, instead of
+      // directly from the source.
+      print_preview_extraction_delegate_->Extract(
+          is_pdf,
+          base::BindOnce(&AIChatTabHelper::OnExtractPrintPreviewContentComplete,
+                         weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
     } else {
-      FetchPageContent(web_contents(), invalidation_token, std::move(callback));
+      // Run print preview extraction after load
+      if (pending_get_page_content_callback_) {
+        std::move(pending_get_page_content_callback_).Run("", false, "");
+      }
+      pending_get_page_content_callback_ = std::move(callback);
     }
+  } else {
+    page_content_fetcher_delegate_->FetchPageContent(
+        invalidation_token,
+        base::BindOnce(&AIChatTabHelper::OnFetchPageContentComplete,
+                       weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
   }
 }
 
-void AIChatTabHelper::PrintPreviewFallback(GetPageContentCallback callback) {
-  pending_print_preview_requests_[pending_navigation_id_] = std::move(callback);
-  if (IsPdf(web_contents())) {
-    // Print preview will be requested in OnPDFA11yInfoLoaded.
-    CheckPDFA11yTree();
-  } else if (is_page_loaded_) {
-    NotifyPrintPreviewRequested(false);
+void AIChatTabHelper::OnFetchPageContentComplete(
+    ConversationHandler::GetPageContentCallback callback,
+    std::string content,
+    bool is_video,
+    std::string invalidation_token) {
+  base::TrimWhitespaceASCII(content, base::TRIM_ALL, &content);
+  if (!did_retry_get_page_content_after_page_load_ && content.empty() &&
+      !is_video) {
+    // Only try this once
+    did_retry_get_page_content_after_page_load_ = true;
+    DVLOG(1) << __func__ << "empty content, will retry once, is_page_loaded_="
+             << is_page_loaded_;
+    if (!is_page_loaded_) {
+      // Retry after page is loaded, including possible print preview fallback
+      if (pending_get_page_content_callback_) {
+        std::move(pending_get_page_content_callback_).Run("", false, "");
+      }
+      pending_get_page_content_callback_ = std::move(callback);
+      return;
+    }
+    if (is_page_loaded_) {
+      // Fallback to print preview extraction
+      print_preview_extraction_delegate_->Extract(
+          IsPdf(web_contents()),
+          base::BindOnce(&AIChatTabHelper::OnExtractPrintPreviewContentComplete,
+                         weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
+    }
```

Comment:
else

Reply:
This wasn't quite an either/or but it's changed a lot now anyway